File: ds_and_algo/implementations/javascript/algorithms/searching/binary-search.js
/**
 *
 * @description - Performs an iterative binary search on a sorted array.
 * @param {Array<number>} arr - The sorted array to search in (must be sorted in ascending order).
 * @param {number} target - The value to search for in the array.
 * @returns {number} - The index of the target value if found, otherwise -1.
 * Time Complexity: O(log n) (worst case) or O(1) (best-case if target element is at the middle index).
 * Space Complexity: O(1). Variables are updated in-place.
 */
function binarySearchIter(arr, target) {
  // Step 1:
  let left = 0;
  let right = arr.length - 1;
  // Step 2:
  while (left <= right) {
    // 2.1:
    const mid = Math.floor((left + right) / 2);
    // 2.2:
    if (arr[mid] === target) {
      return mid; // Element found at index mid.
      // 2.3:
    } else if (arr[mid] < target) {
      left = mid + 1; // Search in the right sub-array.
      // 2.4:
    } else {
      right = mid - 1; // Search in the left sub-array.
    }
  }
  // Step 3:
  return -1; // Element not found.
}

/**
 *
 * @description - Performs a recursive binary search on a sorted array.
 * @param {Array<number>} arr - The sorted array to search in (must be sorted in ascending order).
 * @param {number} target - The value to search for in the array.
 * @param {number} left - The left sub-array.
 * @param {number} right - The right sub-array.
 * @returns {number} - The index of the target value if found, otherwise -1.
 * Time Complexity: O(log n) (worst case) or O(1) (best-case if target element is at the middle index).
 * Space Complexity: O(log n). Due to the depth of the recursive call stack.
 */
function binarySearchRec(arr, target, left = 0, right = arr.length - 1) {
  // Base case (target not found):
  if (left > right) return -1;

  // Calculate the middle index:
  const mid = Math.floor((left + right) / 2);

  // Step 1.1 (target found at index mid):
  if (arr[mid] === target) {
    return mid;
  }
  // Step 1.2:
  if (arr[mid] < target) {
    // Step 2 (conquer/search in the right sub-array):
    return binarySearchRec(arr, target, mid + 1, right);
  }
  // Step 2 (conquer/search in the left sub-array):
  return binarySearchRec(arr, target, left, mid - 1);
}

// Example usage:
if (import.meta.url.endsWith(process.argv[1])) {
  // Iterative:
  const arr = [1, 2, 3, 4, 5];
  let index = binarySearchIter(arr, 4);
  console.log("Iterative Approach:", index); // 3

  // Recursive:
  index = binarySearchRec(arr, 4);
  console.log("Recursive Approach:", index); // 3
}

File: ds_and_algo/implementations/python/algorithms/searching/binary_search.py
def binary_search_iter(arr: list[int], target: int) -> int:
    """
    Performs an iterative binary search on a sorted array.

    Args:
        arr: A sorted list of integers.
        target: The integer value to search for.

    Returns:
        The index of the target element if found, otherwise -1.

    Time complexity: O(log n).
    Space complexity: O(1) for iterative.
    """

    # Step 1:
    left = 0
    right = len(arr) - 1
    # Step 2:
    while left <= right:
        # 2.1:
        mid = (left + right) // 2
        # 2.2:
        if arr[mid] == target:
            return mid  # Element found at index mid.
        # 2.3:
        elif arr[mid] < target:
            left = mid + 1  # Search in the right sub-array.
        # 2.4:
        else:
            right = mid - 1  # Search in the left sub-array.
    # Step 3:
    return -1  # Element not found.


def binary_search_rec(
    arr: list[int],
    target: int,
    left: int = 0,
    right: int | None = None,
) -> int:
    """
    Performs a recursive binary search on a sorted array.

    Args:
        arr: A sorted list of integers.
        target: The integer value to search for.
        left: The leftmost index of the sub-array to search in.
        right: The rightmost index of the sub-array to search in.

    Returns:
        The index of the target element if found, otherwise -1.

    Time complexity: O(log n).
    Space complexity: O(log n) for recursive due to the call stack.
    """

    # Initialize the right sub-array index:
    if right is None:
        right = len(arr) - 1

    # Base case (target not found):
    if left > right:
        return -1

    # Calculate the middle index:
    mid = (left + right) // 2

    # Step 1.1 (target found at index mid):
    if arr[mid] == target:
        return mid
    # Step 1.2:
    if arr[mid] < target:
        # Step 2 (conquer/search in the right sub-array):
        return binary_search_rec(arr, target, mid + 1, right)
    # Step 2 (conquer/search in the left sub-array):
    return binary_search_rec(arr, target, left, mid - 1)


# Example usage:
if __name__ == "__main__":
    # Iterative:
    arr = [1, 2, 3, 4, 5]
    index = binary_search_iter(arr, 4)
    print("Iterative Approach:", index)  # 3

    # Recursive:
    index = binary_search_rec(arr, 4)
    print("Recursive Approach:", index)  # 3
